### DISCUSSION TITLE
How to embed an HTML file in a Quarto Extension

### DISCUSSION THREAD

**Discussion by jmgirard** [quarto-dev/quarto-cli#8994] [Q&A]
I am trying to get the PDF.js library to work for my "embedpdf" quarto extension because that seems to be one of the only ways to embed PDFs in HTML files that will work on mobile devices. However, I seem to need to include the `/web/viewer.html` (and other supporting files) in the extension so that I can embed an iframe in the quarto-rendered HTML file that points to `viewer.html` with some GET codes. However, I'm having a hard time getting this to work. I tried `quarto.doc.add_format_resource()` but Lua doesn't seem to find it and gives a 404 error. Is it possible to include an HTML file like this in my extension?

**Comment by jmgirard** [comment]
I was able to get this working for local use but still failing on GitHub Pages.

**Comment by cderv** [comment]
Are you able to share full example of what you tried ? If it works locally, it is odd it is falling on publishing, unless the `quarto publish` command misses the resource maybe 🤔

**Reply by jmgirard** [reply]
It's a bit complicated, so I apologize for the challenge of making a MWE. But I'll describe below.

**Reply by cderv** [reply]
Even if not a Minimal one, just being able to have a look in your extension, or what you tried to deployed can help us.

**Reply by jmgirard** [reply]
Source: https://github.com/jmgirard/embedpdf

Goal: I want to embed `dummy.pdf` in `example.html` (created by `example.qmd`) in a way that works for mobile

Working locally: 

- Extract PDF.js files to the `pdfjs` folder in same place as `example.qmd`
- Use LUA to add `<div><iframe src="/pdfjs/web/viewer.html?file=../../dummy.pdf"></iframe></div>` to `example.html`

https://github.com/jmgirard/embedpdf/blob/24f00671caf8bca74d91dbe9d253e1830b64b224/_extensions/embedpdf/embedpdf.lua#L37

**Reply by jmgirard** [reply]
The problem seems to be that the `../../dummy.pdf` relative link works locally but not on GHP.

**Reply by jmgirard** [reply]
I'm also not sure how to tell Quarto to include that `pdfjs` folder in the extension, in an `embed-resources: true` file, etc.

**Reply by jmgirard** [reply]
It seems unlikely that a self-contained HTML file can include a separate HTML file that it embeds in itself via iframe. So maybe the solution would be to just instruct users of the extension how to host `viewer.html` separately on their server...

**Reply by mcanouil** [reply]
FYI:

- https://github.com/quarto-dev/quarto-cli/discussions/7918

**Reply by jmgirard** [reply]
@mcanouil Yes, my extension already supports the `<embed>` and `<object>` tags. However, these don't work on (most) mobile devices. So I'm trying to get PDF.js (which does work on mobile) to work with Quarto as a backup to fall to when the former fail.
